tla/counterfeit-cycles/counterfeit_cycles.tla

```diff
@@ -0,0 +1,293 @@
+(* Type checks                                                             *)
+(***************************************************************************)
+
+TransferMessage ==
+  \* A message to transfer cycles between subnets.
+  [type : {"transfer"}, from : SUBNETS, to : SUBNETS, amount : Nat \ {0}]
+
+\* The set of all possible messages that are sent/received by subnets.
+SubnetMessages ==
+  \* Transferring cycles from one subnet to another.
+  TransferMessage
+    \union
+    
+  \* Approving a cycles transfer.
+  \* This message is sent by the ledger to indicate that the `amount` sent is
+  \* now finalized.
+  [type : {"approve"}, from : SUBNETS, to : SUBNETS, amount : Nat \ {0}]  
+
+LedgerMessages ==
+  \* The set of all possible messages that can be received by the Ledger.
+  TransferMessage
+
+SubnetsTypeOK ==
+  /\ \A i \in DOMAIN subnetMsgs: {subnetMsgs[i]} \subseteq SubnetMessages
+  /\ subnets \in [SUBNETS -> [
+        balance: Nat,
+        unfinalized: [SUBNETS -> Nat],
+        msgsToLedger: Seq(LedgerMessages),
+        honest: {TRUE, FALSE}
+     ]]
+
+LedgerTypesOK ==
+  \* Type-correctness invariant.
+  /\ DOMAIN ledger.balances = SUBNETS
+  /\ \A s \in SUBNETS: ledger.balances[s] >= 0
+  /\ ledger.dishonestSubnets \subseteq SUBNETS
+  /\ \A i \in DOMAIN ledger.msgs: {ledger.msgs[i]} \subseteq LedgerMessages
+  
+TypesOK ==
+  \* Type-correctness invariant.
+  /\ LedgerTypesOK
+  /\ SubnetsTypeOK
+
+-----------------------------------------------------------------------------
+(***************************************************************************)
+(* Subnet actions                                                          *)
+(***************************************************************************)
+
+SubnetSendTransfer ==
+  (* Honestly sends a transfers cycles from one subnet to another. *)
+  
+  \* Choose two subnets.
+  \E sender, receiver \in SUBNETS: sender /= receiver
+  /\ subnets[sender].balance > 0
+  
+  \* Choose an amount that's within the subnet's balance.
+  /\ \E amount \in 1..subnets[sender].balance:
+  
+  \* Send a transfer to the receiver.
+  /\ subnetMsgs' = Append(subnetMsgs, [
+        type |-> "transfer",
+        from |-> sender,
+        to |-> receiver,
+        amount |-> amount
+     ])
+     
+  \* Subtract amount from sender's balance.
+  /\ subnets' = [subnets EXCEPT ![sender].balance = @ - amount]
+  
+  /\ UNCHANGED<<ledger>>
+
+SubnetDishonestSendTransfer ==
+  (* Maliciously transfers more cycles than the subnet's own balance. *)
+  
+  \E sender, receiver \in SUBNETS: sender /= receiver
+
+  \* To keep the state space bounded, a subnet can make a dishonest
+  \* transfer at most once.
+  /\ subnets[sender].honest
+  
+  \* Send a transfer to the receiver.
+  /\ subnetMsgs' = Append(subnetMsgs, [
+        type |-> "transfer",
+        from |-> sender,
+        to |-> receiver,
+        \* Choose an amount that's greater than the balance.
```

Comment:
I know we discussed potentially improving this but I don't have a really good suggestion for improvement here :( Obviously you need to limit the state space so you need to pick an amount from some finite set. You could pick an amount from `1..MAX_COUNTERFEIT_CYCLES` or something like that, that might be a bit nicer.

Reply:
I agree. Interestingly enough, when I did this, I discovered a behavior that breaks one of the invariants. I'll follow up on this :)

Reply:
OK, you got my attention :) could it be an off-by-1 error somewhere? I guess that would be easy to miss with a +10 increase.

Reply:
The issue we're running into here is that the model assumes that whether or not a transfer is "honest" is binary (i.e. a transfer is either honest or dishonest). This isn't exactly true. For example, a subnet with a balance of 2 can dishonestly try transfer 3 cycles, but as it's doing so it receives 1 cycle, and therefore the transfer is "honest" as there are enough cycles available.

When we add +10, because that's more than the total supply, the example above never occurs. I think I need to think a bit harder about what the properties and invariants that we want to maintain, and what it really means for something to be "honest" - maybe quantifying what's "honest" isn't even relevant and can be removed 🤷‍♂️

Reply:
OK, that's interesting, I haven't thought about that case. I'd say that the transfer from your example is dishonest simply by definition, because an honest subnet would not do such a transfer. But I see that it violates the property that dishonest subnets are eventually detected. As you say maybe we can live without such a property, or maybe we can change the protocol to still detect such subnets. 

But I can only think of quite complicated solutions, something like: the ledger signs current balances together with a ledger time, and then the subnets provide their latest known signature of their balance with outgoing transfers, which the receiving subnet forwards to the ledger.  Eventually the ledger could do a reconciliation of all the records. But it's really complicated (and the tECDSA signatures are quite possibly too slow for this, and certified variables don't work in replicated mode), it would require unbounded storage on the ledger (unless we can come up with some garbage collection scheme). It also wouldn't detect the case where dishonest subnets send cycles to each other in invalid ways, but I think it's safe to say we will never be able to detect this reliably.

Reply:
Right, for simplicity, I think the actual detection of whether or not a subnet behaved honestly isn't the goal here. The goal is to detect and stop the proliferation of counterfeit cycles.

I have updated the code now to address this, where I added `MAX_COUNTERFEIT_CYCLES` constant and updated the invariants accordingly.